docs/architecture/adr-071-proposer-based-timestamps.md

```diff
@@ -0,0 +1,458 @@
+	Height     int64       `json:"height"`
+	Round      int32       `json:"round"`
+	BlockID    BlockID     `json:"block_id"`
+	Signatures []CommitSig `json:"signatures"`
+}
+```
+
+`Commit` will be updated to contain the following fields: 
+
+```go
+type Commit struct {
+	Height     int64       `json:"height"`
+	Round      int32       `json:"round"`
+	Timestamp   time.Time   `json:"timestamp"` 
+	BlockID    BlockID     `json:"block_id"`
+	Signatures []CommitSig `json:"signatures"`
+}
+```
+
+### Changes to `Vote` messages
+
+`Precommit` and `Prevote` messages use a common [Vote struct](https://github.com/tendermint/tendermint/blob/a419f4df76fe4aed668a6c74696deabb9fe73211/types/vote.go#L50).
+This struct currently contains a timestamp.
+This timestamp is set using the [voteTime](https://github.com/tendermint/tendermint/blob/e8013281281985e3ada7819f42502b09623d24a0/internal/consensus/state.go#L2241) function and therefore vote times correspond to the validator’s local time.
+For precommits, this timestamp is used to construct the [CommitSig that is included in the block in the LastCommit](https://github.com/tendermint/tendermint/blob/e8013281281985e3ada7819f42502b09623d24a0/types/block.go#L754) field.
+For prevotes, this field is unused.
+Proposer-based timestamps will use the [RoundState.Proposal](https://github.com/tendermint/tendermint/blob/c3ae6f5b58e07b29c62bfdc5715b6bf8ae5ee951/internal/consensus/types/round_state.go#L76) timestamp to construct the `signedBytes` `CommitSig`.
+This timestamp is therefore no longer useful and will be dropped.
+
+`Vote` currently contains the following fields:
+
+```go
+type Vote struct {
+	Type             tmproto.SignedMsgType `json:"type"`
+	Height           int64                 `json:"height"`
+	Round            int32                 `json:"round"`    
+	BlockID          BlockID               `json:"block_id"` // zero if vote is nil.
+	Timestamp        time.Time             `json:"timestamp"`
+	ValidatorAddress Address               `json:"validator_address"`
+	ValidatorIndex   int32                 `json:"validator_index"`
+	Signature        []byte                `json:"signature"`
+}
+```
+
+`Vote` will be changed to contain the following fields:
+
+```go
+type Vote struct {
+	Type             tmproto.SignedMsgType `json:"type"`
+	Height           int64                 `json:"height"`
+	Round            int32                 `json:"round"` // proposal/vote round   
+	ForBlock         bool                  `json:"for_block"` // false if vote is nil.
+	ValidatorIndex   int32                 `json:"validator_index"`
+	Signature        []byte                `json:"signature"`
+}
+```
+
+`BlockID` is also dropped as it is also stored by all validators.
+`ValidatorIndex` is all that is needed to determine `ValidatorAddress` so it is dropped as well.
+
+### New consensus parameters
+
+The proposer-based timestamp specification includes multiple new parameters that must be the same among all validators.
```

Comment:
There's an important distinction here. Params such as the block size are verifiable i.e. I know when a proposer is faulty and proposes a block too big. These params, however, are non-verifiable.  They are used to signal to honest nodes what value to run with but nodes don't validate the actual times that other nodes implement. If a node wants to use a different value, it can but this is theoretically classified as a faulty node whereby greater than 1/3+ _may_ cause the network to halt. The consensus mechanism should ensure that changes to these parameters are atomic. These could only be non atomic if there was a fork (which requires greater than 1/3) or the application was non-deterministic (which could halt the system).

Reply:
To expand a bit on @cmwaters explanation: proposer's place a hash of their 'consensus params' into the block in the `consensus_hash` field. This hash is checked when a node [validates a block](https://github.com/tendermint/tendermint/blob/18784f7fe0fa316994dd94cbd84573d87d7958d4/state/validation.go#L60). Validators check that the consensus hash matches their locally defined consensus hash and reject the block if there is a mismatch. This only verifies that what the proposer _reports_ as its consensus hash matches the validator's consensus hash. A malicious proposer could certainly report a hash of parameters that are different from the parameters it actually is acting with. I.e., it could include '20mb' as the max block size but then only prevote blocks that are 10mb or less. It's a mechanism to ensure correct validators agree but malicious ones can certainly bypass it and behave differently. 

> what can go wrong if validators are updated with different values for these parameters, and the update isn't atomic? Would changing these values potentially require a hard fork?

If correct validators are updated non-atomically, then there will be a split: updated validators will not prevote blocks proposed by non-updated validators and vice-versa. These types of changes require a coordinated upgrade: All nodes must agree on a height to perform the upgrade.